internal/garbagecollector/garbageCollector.go

```diff
@@ -0,0 +1,95 @@
+package garbagecollector
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/jinzhu/gorm"
+	clusterPkg "github.com/openshift/assisted-service/internal/cluster"
+	"github.com/openshift/assisted-service/internal/host"
+	"github.com/openshift/assisted-service/pkg/leader"
+	"github.com/openshift/assisted-service/pkg/s3wrapper"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	DeletedUnregisteredAfter time.Duration `envconfig:"DELETED_UNREGISTERED_AFTER" default:"168h"` // 7d
+	DeregisterInactiveAfter  time.Duration `envconfig:"DELETED_INACTIVE_AFTER" default:"720h"`     // 20d
+	MaxGCClustersPerInterval int           `envconfig:"MAX_GC_CLUSTERS_PER_INTERVAL" default:"100"`
+}
+
+type GarbageCollectors interface {
+	PermanentClustersDeletion(ctx context.Context, olderThan strfmt.DateTime, objectHandler s3wrapper.API) error
+	DeregisterClusterInternal(ctx context.Context, params installer.DeregisterClusterParams) error
+}
+
+func NewGarbageCollectors(
+	Config Config,
+	db *gorm.DB,
+	log logrus.FieldLogger,
+	hostApi host.API,
+	clusterApi clusterPkg.API,
+	objectHandler s3wrapper.API,
+	leaderElector leader.Leader,
+
+) *garbageCollector {
+	return &garbageCollector{
+		Config:        Config,
+		db:            db,
+		log:           log,
+		hostApi:       hostApi,
+		clusterApi:    clusterApi,
+		objectHandler: objectHandler,
+		leaderElector: leaderElector,
+	}
+}
+
+type garbageCollector struct {
+	Config
+	db            *gorm.DB
+	log           logrus.FieldLogger
+	hostApi       host.API
+	clusterApi    clusterPkg.API
+	objectHandler s3wrapper.API
+	leaderElector leader.Leader
+}
+
+func (g garbageCollector) DeregisterInactiveClusters() {
+	if !g.leaderElector.IsLeader() {
+		g.log.Debugf("Not a leader, exiting periodic inactive clusters deregistration")
+		return
+	}
+
+	olderThan := strfmt.DateTime(time.Now().Add(-g.Config.DeregisterInactiveAfter))
+	g.log.Infof("Deregistering %d clusters that were inactive since %s", g.MaxGCClustersPerInterval, olderThan)
```

Comment:
the message is incorrect, you don't know if 100 or less clusters will be deleted